### ISSUE TITLE
Loss tends to increase when the gradient accumulation steps are increased while using the SFTTrainer

### ISSUE THREAD

**Issue by Asnegha** [huggingface/trl#2604]
### Reproduction

Llama model is instruction fine tuned using SFTTrainer. When I make gradient accumulation from 1 to 4 the loss just gets increased by same factor. The blue graph belongs to training loss with gradient accumulation step as 1 and orange graph belongs to training loss with gradient accumulation step 2.

![Image](https://github.com/user-attachments/assets/bc94eff6-0313-426d-9807-36fd78ae48c7)

```python
    training_arguments = TrainingArguments(
    save_strategy="steps",
    save_steps=4000,
    output_dir=args.output_dir,
    num_train_epochs=2,
    gradient_accumulation_steps = 4,
    per_device_train_batch_size= 8,
    per_device_eval_batch_size=4,
    lr_scheduler_type ='cosine',
    learning_rate= 0.01,
)
    trainer = SFTTrainer(
    model=model,
    train_dataset=train_dataset,
    peft_config=peft_config,
    dataset_text_field= "prompt",
    tokenizer=tokenizer,
    args=training_arguments,
    packing= False,
    max_seq_length= 1024
    )
```

### System Info

OS type and version: Linux
Python:  3.10.14 
PyTorch:   2.4.0+cu124 
TRL:  0.9.6
Transformers:  4.47.0 

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
Thanks for reporting. Do you get the results with the latest transformers release? Can you provide a full minimal code to reproduce?

**Comment by albertvillanova** [comment]
Hey @Asnegha, sorry for the slow reply — we're catching up on the backlog. 

This is the gradient-accumulation loss-normalization bug from that era: TRL 0.9.6's `SFTTrainer` predated the `num_items_in_batch` plumbing, so paired with transformers 4.47 the SFT loss wasn't normalized correctly and scaled with the accumulation factor. Current TRL reduces the loss as `sum / num_items_in_batch`, which is gradient-accumulation-correct (see https://github.com/huggingface/trl/blob/fceadc75c81bd103a6cf7f870eafddf1106b6065/trl/trainer/sft_trainer.py#L231-L236), so upgrading should make GA=1 and GA=4 give comparable loss curves. 

Closing as fixed — happy to re-open if you still see it on a current TRL/transformers with a full script and `trl env` output. (Unrelated aside: `learning_rate=0.01` is very high for LoRA SFT — you likely want ~1e-4–2e-4.)
